esrally/props.py

```diff
@@ -0,0 +1,113 @@
+# Licensed to Elasticsearch B.V. under one or more contributor
+# license agreements. See the NOTICE file distributed with
+# this work for additional information regarding copyright
+# ownership. Elasticsearch B.V. licenses this file to you under
+# the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# 	http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from enum import Enum, auto
+
+try:
+    from enum import EnumType
+except ImportError:
+    from enum import EnumMeta as EnumType
+
+try:
+    from enum import member
+except ImportError:
+
+    def member(enum):
+        return enum
+
+
+class PropEnumType(EnumType):
+    __iter__ = None
+
+
+class PropEnum(str, Enum, metaclass=PropEnumType):
+    def __new__(cls, value):
+        if isinstance(value, PropEnumType):
+            return value
+        elif isinstance(value, str):
+            value = f"{cls.__qualname__.lower()}.{value}"
+            member = str.__new__(cls, value)
+            member._value_ = value
+            return member
+        else:
+            raise TypeError(f"{value!r} is not a string or enum")
```

Comment:
I wrote down my best understanding in the comments. Hope they clarify things a little. But note that they may have errors because the enum is too hard to understand. I'm not confident until I test the code by unit test cases later.